### ISSUE TITLE
Devise and enforce a grammar for project names at "obi new" time

### ISSUE THREAD

**Issue by tomjakubowski** [Oblong/obi#48]
A template wants to use `{{ project_name }}` in one of its jinja2 templates to scaffold some C++ code.  If the project name contains "weird" characters (unsuitable for a C++ identifier, say), havoc ensures when the user tries to build the project.   It would be nice if templates could make some basic assumptions about the project name.

**Comment by EchoAbstract** [comment]
I did a hack to work around this in the obi-growroom template, see https://github.com/Oblong/obi-growroom/blob/master/growroom.py#L65-L69

It's not great, since it's possible to get duplicate symbols here, but it works for growroom, since it ends up only being used for binding stuff.

**Comment by sandover** [comment]
What would it be like if obi enforced total project name simplicity, along the lines of [a-zA-Z]+ ?  Maybe allowing digits as well.

**Comment by EchoAbstract** [comment]
@Gobiner can we at least add non-leading digits here?  Also, this needs to be featured prominently in the README, since it's non-intuative that my project name has to conform to the restrictions.

**Comment by jshrake** [comment]
I think we should enforce the following: `(letter | '_') (letter | digit | '_')*`

My attempt at the regex: `^[a-zA-Z_]\w*$`

**Comment by sandover** [comment]
does that regex allow for digits @jshrake?

**Comment by jshrake** [comment]
If memory serves, `\w` is shorthand for `[a-zA-Z0-9_]`. We could just write that out explicitly.

**Comment by jshrake** [comment]
We should probably allow `-`in the name, so `^[a-zA-Z_][\w-]*$` or ^[a-zA-Z_][a-zA-Z0-9_-]*$`

**Comment by sandover** [comment]
i'm excited  we are finally getting the upper hand on underscores

**Comment by EchoAbstract** [comment]
@jshrake if we allow `-` then it is no longer a valid C++ identifier, so we'd need to do something like my hack in obi-growroom to generate a valid identifier.  (I'm a proponent of this approach, but it's got pitfalls)

**Comment by sandover** [comment]
What are the pros/cons of having project names also be valid C++ identifiers?  It seems appealing but I'm wondering about the specifics.

**Comment by EchoAbstract** [comment]
We use it in greenhouse to build the binding symbol, the binding C++ source files, and the binding JS namespace.  It's _nicer_ than just using the word binding, but we could work around it.  The nice part of @jshrake's regex above is that we can easily transform the project name into an identifier.

**Comment by tomjakubowski** [comment]
My preference is for `^[a-zA-Z][a-zA-Z0-9_-]*$` (obi should not allow leading underscores, which could cause a user to inadvertedly step on "reserved identifier" turf with a project name like "_Foo" or "__bar"). obi can provide a little jinja2 helper to replace `-` with `_` for the purposes of C++ identifier generation.

**Comment by Gobiner** [comment]
Here's a list of the ways the obi project name is used in templates today:
* as the name of the new directory
    * `/Users/arice/{{project_name}}/project.yaml`
* as part of growroom generated filenames
    * `"${PROJECT_SOURCE_DIR}/src/generated/${PROJECT_NAME}_seabed_binding.cpp"`
* as a string value in the project.yaml
    * `name: '{{project_name}}'`
* as a symbol in CMakeLists.txt
    * `project({{project_name}})`
    * `add_executable({{project_name}}`
* as a debian package name in debian/control
    * `Source: oblong-{{project_name}}-gs{{g_speak_version}}x1`
* as part of a C++ identifier in growroom/seabed
    * `js_executor -> AddJSBinding (Bind_{{project_name}}_seabed_In_Global);`
* as pool names in config.protein
    * `chief-pool: {{project_name}}`
    * `heartbeat-pool: tcp://localhost/{{project_name}}-pulse`
* as part of a regex in `set_gspeak.sh`
    * `sed -i.bak  "s/oblong-{{project_name}}-gs[0-9.]*x[0-9]*/oblong-{{project_name}}-gs${gspeak_version}x$astor_major/g"`

We probably want to keep in mind these uses while choosing a regex.

**Comment by sandover** [comment]
- I would like to avoid the hyphen/underscore thumb wars of `packagey-thing-narb_1_flarg-bar_norf-1-2-3.deb`, which we cannot win
- I would like to consolidate Oblong's practice in the file system (names of folders and names of executables) around hyphens only; but since we cannot always win with hyphens, the next best thing is to do with out and stick with:
- short sweet project names
- I would like to keep obi's logic simple
- I would like to keep the growroom / code gen logic simple

For all those reasons, `^[a-zA-Z_]\w*$` seems clean.

**Comment by sandover** [comment]
@Gobiner thanks for grounding us in the facts!  That helps.

**Comment by tomjakubowski** [comment]
Three cheers for @Gobiner grounding us in reality.

**Comment by dankegel** [comment]
set-gspeak.sh should be gone if it isn't already; ob-set-defaults has that duty now.

There are valid reasons to allow dashes in project names (with the proviso
that - would have to map to _ inside C++ identifiers).
For instance, if you're using obi to regenerate an existing project
that has a dash in the project name :-)

When I had to do this recently, I worked around the check by just hacking obi's source
to remove the restriction.  Might be good to add a --allow-dashes option to keep
cantankerous users from shooting themselves in the foot.

Underscores in project names should be forbidden forever IMHO, so the
mapping could be reversible.
